### DISCUSSION TITLE
type caster that turns values into python exceptions?

### DISCUSSION THREAD

**Discussion by ghorn** [wjakob/nanobind#755] [Q&A] [answered]
We use `std::expected` for error handling in C++. In python, we never handle the unexpected case - for all `unexpected` values we just throw exceptions. I tried writing a type caster that did this automatically by throwing an exception in `from_cpp`, but I observed that it calls `std::terminate` and does not simply raise this into a python exception as I expected.

This seems to be intentional behavior - from reading the source and from the answer here https://github.com/wjakob/nanobind/discussions/272

Should I abandon this approach? Perhaps the better approach would be to do this in the user binding code, not the type caster code for `std::expected`.

**Comment by wjakob** [comment, marked as answer]
Indeed, type casters can't raise by design. ``std::expected`` as a C++23 feature is also beyond what can be handled in nanobind, which is a C++17 library. 

You could either handle this on the Python interface or fork nanobind to handle `std::expected` specially. However, I would be unlikely to merge such a feature here, in case that is relevant.

**Comment by oremanj** [comment]
It *is* possible for `from_cpp` to raise an exception, but it has to be done in the Python way: call `PyErr_SetXyz` or `PyErr_Format`, then return nullptr.

**Reply by wjakob** [reply]
Right, that's a good point!
